optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -181,4 +187,56 @@ public boolean isFieldADeepCloneProperty(Field field, Class<?> owningClass) {
         }
         return false;
     }
+
+    /**
+     * @return never null
+     */
+    public Set<Class<?>> getDeepClonedTypeArguments(Type genericType) {
+        // Check the generic type arguments of the field.
+        // Yes, it is possible for fields and methods, but not instances!
+        if (genericType instanceof ParameterizedType) {
+            Set<Class<?>> deepClonedTypeArguments = new HashSet<>();
+            ParameterizedType parameterizedType = (ParameterizedType) genericType;
+            for (Type actualTypeArgument : parameterizedType.getActualTypeArguments()) {
+                if (actualTypeArgument instanceof Class
+                        && isClassDeepCloned((Class) actualTypeArgument)) {
+                    deepClonedTypeArguments.add((Class) actualTypeArgument);
+                }
+                deepClonedTypeArguments.addAll(getDeepClonedTypeArguments(actualTypeArgument));
+            }
+            return deepClonedTypeArguments;
+        }
+        return Collections.emptySet();
```

Comment:
Personally, I'd invert the logic here. First get the simple thing out of the way, then deal with the more complex thing.